File: backend/app/storage/providers/s3_provider.py
import boto3
from botocore.exceptions import ClientError
from typing import Dict, Any
from app.config import settings
from app.storage.providers.storage_provider import StorageProvider

class S3StorageProvider(StorageProvider):
    def __init__(self):
        client_kwargs = {
            "region_name": settings.AWS_REGION,
        }
        
        # Inject access keys if provided
        if settings.AWS_ACCESS_KEY_ID:
            client_kwargs["aws_access_key_id"] = settings.AWS_ACCESS_KEY_ID
        if settings.AWS_SECRET_ACCESS_KEY:
            client_kwargs["aws_secret_access_key"] = settings.AWS_SECRET_ACCESS_KEY
            
        # Support emulator URLs (LocalStack/MinIO)
        if settings.AWS_S3_ENDPOINT_URL:
            client_kwargs["endpoint_url"] = settings.AWS_S3_ENDPOINT_URL
            
        self.s3_client = boto3.client("s3", **client_kwargs)
        self.bucket_name = settings.AWS_S3_BUCKET_NAME

    def upload_file(self, tenant_id: str, category: str, filename: str, content: bytes) -> str:
        s3_key = f"tenants/{tenant_id}/{category}/{filename}"
        try:
            self.s3_client.put_object(
                Bucket=self.bucket_name,
                Key=s3_key,
                Body=content
            )
            return s3_key
        except ClientError as e:
            raise RuntimeError(f"Failed to upload object to S3 bucket: {e}")

    def download_file(self, key: str) -> bytes:
        try:
            response = self.s3_client.get_object(
                Bucket=self.bucket_name,
                Key=key
            )
            return response["Body"].read()
        except ClientError as e:
            raise FileNotFoundError(f"Failed to retrieve S3 object '{key}': {e}")

    def delete_file(self, key: str) -> bool:
        try:
            self.s3_client.delete_object(
                Bucket=self.bucket_name,
                Key=key
            )
            return True
        except ClientError:
            return False

    def get_presigned_download_url(self, key: str, expires_in: int = 3600) -> str:
        try:
            url = self.s3_client.generate_presigned_url(
                ClientMethod="get_object",
                Params={
                    "Bucket": self.bucket_name,
                    "Key": key
                },
                ExpiresIn=expires_in
            )
            return url
        except ClientError as e:
            raise RuntimeError(f"Failed to generate pre-signed S3 download link: {e}")

    def get_presigned_upload_url(self, tenant_id: str, category: str, filename: str, expires_in: int = 3600) -> Dict[str, Any]:
        s3_key = f"tenants/{tenant_id}/{category}/{filename}"
        try:
            response = self.s3_client.generate_presigned_post(
                Bucket=self.bucket_name,
                Key=s3_key,
                ExpiresIn=expires_in
            )
            return response
        except ClientError as e:
            raise RuntimeError(f"Failed to generate pre-signed S3 upload payload: {e}")

File: backend/src/files/storage/s3.provider.ts
import {
  S3Client,
  PutObjectCommand,
  GetObjectCommand,
  DeleteObjectCommand,
} from '@aws-sdk/client-s3';
import { getSignedUrl } from '@aws-sdk/s3-request-presigner';
import { StorageProvider } from './storage.interface';

export class S3StorageProvider implements StorageProvider {
  private s3: S3Client;
  public readonly bucketName: string;

  constructor() {
    const config: any = {
      region: process.env.AWS_REGION || 'ap-south-1',
    };

    if (process.env.AWS_ACCESS_KEY_ID) {
      config.credentials = {
        accessKeyId: process.env.AWS_ACCESS_KEY_ID,
        secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY,
      };
    }

    if (process.env.AWS_S3_ENDPOINT_URL) {
      config.endpoint = process.env.AWS_S3_ENDPOINT_URL;
      config.forcePathStyle = true;
    }

    this.s3 = new S3Client(config);
    this.bucketName = process.env.AWS_S3_BUCKET_NAME || 'legaldraft-bucket';
  }

  async uploadFile(
    tenantId: string,
    category: string,
    filename: string,
    content: Buffer,
  ): Promise<string> {
    const s3Key = `tenants/${tenantId}/${category}/${filename}`;
    await this.s3.send(
      new PutObjectCommand({
        Bucket: this.bucketName,
        Key: s3Key,
        Body: content,
      }),
    );
    return s3Key;
  }

  async downloadFile(key: string): Promise<Buffer> {
    const response = await this.s3.send(
      new GetObjectCommand({
        Bucket: this.bucketName,
        Key: key,
      }),
    );
    const chunks: Buffer[] = [];
    for await (const chunk of response.Body as any) {
      chunks.push(Buffer.from(chunk));
    }
    return Buffer.concat(chunks);
  }

  async deleteFile(key: string): Promise<boolean> {
    try {
      await this.s3.send(
        new DeleteObjectCommand({
          Bucket: this.bucketName,
          Key: key,
        }),
      );
      return true;
    } catch {
      return false;
    }
  }

  async getPresignedDownloadUrl(
    key: string,
    expiresIn = 3600,
  ): Promise<string> {
    const command = new GetObjectCommand({
      Bucket: this.bucketName,
      Key: key,
    });
    return getSignedUrl(this.s3, command, { expiresIn });
  }

  async getPresignedUploadUrl(
    tenantId: string,
    category: string,
    filename: string,
    expiresIn = 3600,
  ): Promise<{ url: string; fields: Record<string, any> }> {
    const s3Key = `tenants/${tenantId}/${category}/${filename}`;
    const command = new PutObjectCommand({
      Bucket: this.bucketName,
      Key: s3Key,
    });
    const url = await getSignedUrl(this.s3, command, { expiresIn });
    return { url, fields: { key: s3Key, method: 'PUT' } };
  }
}
